docs/CODING_CONVENTIONS.md

```diff
@@ -0,0 +1,260 @@
+
+###### Example
+
+Implementing a non-exhaustive variant is done like this:
+
+```rust
+pub enum Error {
+    PermissionDenied,
+    NotConnected,
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+```
+
+This informs users of the API to include a catch-all arm when matching against the error:
+
+```rust
+match e {
+    PermissionDenied => { /*  */ },
+    NotConnected => { /*  */ },
+    _ => { /*  */ },
+}
+```
+
+This pattern guards against any matches being non-exhaustive in case new error variants are added in the future.
+
+#### Use `amethyst_error::Error` for compositional APIs
+
+APIs which composes results from multiple crates _should_ use `amethyst_error::Error`.
+This is a generic error type which is capable of boxing any error and annotate it with debugging information.
+
+This must be used when defining APIs which composes errors generically, like with traits.
+This is also required when composing errors from other crates, since it is currently the only mechanism available to
+communicate backtraces.
+
+###### Do
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// no dependencies
+mod c {
+    use amethyst_error::Error;
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+###### Don't
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::A(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::B(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// depends on `a` and `b`.
+mod c {
+    use crate::{a, b};
+
+    pub enum Error {
+        A(a::error::Error),
+        B(b::error::Error),
+    }
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+#### Avoid overloading the default `Result`
+
+We _should not_ import a `Result` that overloads the default import.
+Instead, if we are using a `Result` alias, use it through a module (e.g. `io::Result`).
```

Comment:
Alternatively you can import with `as` i.e. `use std::io::Result as IoResult`

Reply:
my biggest problem with this is that you lose the module path, as generally io::Result and IoResult are the same but with `io::Result` you know the concrete type is still `Result` but its the one that comes from `io`, just slightly more clear imho